cdap-app-fabric/src/test/java/io/cdap/cdap/internal/operation/InMemoryOperationRunnerTest.java

```diff
@@ -0,0 +1,109 @@
+import io.cdap.cdap.sourcecontrol.ApplicationManager;
+import io.cdap.cdap.sourcecontrol.operationrunner.InMemorySourceControlOperationRunner;
+import java.time.Instant;
+import java.util.Collections;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class InMemoryOperationRunnerTest {
+
+  private InMemoryOperationRunner runner;
+
+  private static Injector injector;
+  private static final OperationStatePublisher mockPublisher = Mockito.mock(
+      OperationStatePublisher.class);
+  private static final InMemorySourceControlOperationRunner mockScmRunner = Mockito.mock(
+      InMemorySourceControlOperationRunner.class);
+  private static final ApplicationManager mockManager = Mockito.mock(ApplicationManager.class);
+  private static final OperationRunId runId = new OperationRunId("namespace", "run");
+  private static final OperationRun run = OperationRun.builder()
+      .setRunId(runId.getRun())
+      .setStatus(OperationRunStatus.PENDING)
+      .setType(OperationType.PULL_APPS)
+      .setMetadata(
+          new OperationMeta(Collections.emptySet(), Instant.now(), null))
+      .build();
+  private static final OperationRunDetail detail = OperationRunDetail.builder()
+      .setSourceId(AppFabricTestHelper.createSourceId(0))
+      .setRunId(runId)
+      .setRun(run)
+      .setPullAppsRequest(new PullAppsRequest(ImmutableSet.of("1", "2", "3", "4"), null))
+      .build();
+
+  @BeforeClass
+  public static void setupClass() {
+    CConfiguration cconf = CConfiguration.create();
+    injector = AppFabricTestHelper.getInjector(cconf, new AbstractModule() {
+      @Override
+      protected void configure() {
+        bind(OperationStatePublisher.class).toInstance(mockPublisher);
+        bind(InMemorySourceControlOperationRunner.class).toInstance(mockScmRunner);
+        bind(ApplicationManager.class).toInstance(mockManager);
+      }
+    });
+  }
+
+  @Before
+  public void setupTest() {
+    runner = injector.getInstance(InMemoryOperationRunner.class);
+    Mockito.reset(mockPublisher);
+  }
+
+  @Test
+  public void testRun() throws Exception {
+
+    OperationController controller = runner.run(detail);
+    controller.completionFuture().get();
+
+    Mockito.verify(mockPublisher, Mockito.times(1)).publishRunning(runId);
+    Mockito.verify(mockPublisher, Mockito.times(1)).publishSuccess(runId);
+  }
+
+  @Test
+  public void testRunWithError() throws Exception {
+
+    Mockito.doThrow(new RuntimeException("test")).when(mockScmRunner).pull(
+        Mockito.any(), Mockito.any()
+    );
+
+    OperationController controller = runner.run(detail);
+    controller.completionFuture().get();
+
+    Mockito.verify(mockPublisher, Mockito.times(1)).publishRunning(runId);
+    Mockito.verify(mockPublisher, Mockito.times(1)).publishFailed(Mockito.eq(runId), Mockito.any());
+    Mockito.verify(mockPublisher, Mockito.never()).publishSuccess(runId);
+  }
+}
+
```

Comment:
Will add a todo for stop test. I tried to add some delay in mock api call but it did not work as expected and the operation was finishing before stop was being called